### ISSUE TITLE
Add a callback for the safe.directory error handling

### ISSUE THREAD

**Issue by csware** [libgit2/libgit2#6430]
Have a repository with a directory not owned by the current user (or an admin).

The git_repository_open returns GIT_EOWNER and provides a string containing an error message (cf. https://github.com/libgit2/libgit2/commit/af9e00321f06383ccf61f4ca9616716bec0eb2da).

For GUI applications using libgit2 there is no way to get the reported path in an easy way, e.g. for providing an option for adding it to the allow list (is it the gitdir, the worktree, or something else). Git for Windows also performs some normalizations (such as resolving symlinks/mount points/junktions) - this is also necessary for libgit2 (cf. #6379) an d would make application developers reimplement the very same logic.

I suppose a callback is needed to which the path is reported so that it can be used easily for further handling such as allowing.

### Version of libgit2 (release number or SHA1)

1.5.0

**Comment by ethomson** [comment]
Yeah, this is a pain point. Today you'd need to screen-scrape the error message which is obviously disappointing.

In general, this feels like another case of the fact that we need a warnings subsystem that callers can subscribe to. In this case, you should be able to get a structured piece of data that indicates that there was a safe directory error, get the directory in question, and be able to return something that tells libgit2 to continue its handling of the problem (in this case, returning an error) or overriding the problem.

(Maybe even retrying?)

Since these problems can happen at the repository level, this would presumably need to happen globally. Some callback added at the `git_libgits_opts` level.

Let me sketch out an idea and we can see if it works.

**Comment by ethomson** [comment]
@csware see https://github.com/libgit2/libgit2/pull/6485 🤔

**Comment by csware** [comment]
In gerneral, looks good.

Is there a need that the callback can ask to continue or ignore? I acknowledge, that this makes it very flexible, however, also could allow to implement a different logic. Do we need this freedom?

**Comment by ethomson** [comment]
> Maybe a specific struct might be better that using varargs if multiple parameters are needed?! (also in the sense of extensability)

Yeah, that's what was proposed here -- https://github.com/libgit2/libgit2/pull/3722/files -- which I also like. I was hoping that there was an _even simpler_ path forward, rather than creating a bunch of structs. But... I do concede that my PR is simple from _libgit2's point-of-view_ and perhaps not _an end-user's_, while #3722 might be simpler for an end-user. I'll need to play around a bit with both. 😓

**Comment by csware** [comment]
Btw. I just notices that the callback would also need to pass the right path for UNC paths (i.e., prefixed with `prefix)/`)...
